algosdk/atomic_transaction_composer.py

```diff
@@ -0,0 +1,529 @@
+from abc import ABC, abstractmethod
+import base64
+import copy
+from enum import IntEnum
+
+from algosdk import abi, error
+from algosdk.future import transaction
+
+# The first four bytes of an ABI method call return must have this hash
+ABI_RETURN_HASH = b"\x15\x1f\x7c\x75"
+
+
+class AtomicTransactionComposerStatus(IntEnum):
+    # BUILDING indicates that the atomic group is still under construction
+    BUILDING = 0
+
+    # BUILT indicates that the atomic group has been finalized,
+    # but not yet signed.
+    BUILT = 1
+
+    # SIGNED indicates that the atomic group has been finalized and signed,
+    # but not yet submitted to the network.
+    SIGNED = 2
+
+    # SUBMITTED indicates that the atomic group has been finalized, signed,
+    # and submitted to the network.
+    SUBMITTED = 3
+
+    # COMMITTED indicates that the atomic group has been finalized, signed,
+    # submitted, and successfully committed to a block.
+    COMMITTED = 4
+
+
+class AtomicTransactionComposer:
+    """
+    Constructs an atomic transaction group which may contain a combination of
+    Transactions and ABI Method calls.
+
+    Args:
+        status (AtomicTransactionComposerStatus): IntEnum representing the current state of the composer
+        method_dict (dict): dictionary of an index in the transaction list to a Method object
+        txn_list (list[TransactionWithSigner]): list of transactions with signers
+        signed_txns (list[SignedTransaction]): list of signed transactions
+        tx_ids (list[str]): list of individual transaction IDs in this atomic group
+    """
+
+    MAX_GROUP_SIZE = 16
+
+    def __init__(self) -> None:
+        self.status = AtomicTransactionComposerStatus.BUILDING
+        self.method_dict = {}
+        self.txn_list = []
+        self.signed_txns = []
+        self.tx_ids = []
+
+    def get_status(self):
+        """
+        Returns the status of this composer's transaction group.
+        """
+        return self.status
+
+    def get_tx_count(self):
+        """
+        Returns the number of transactions currently in this atomic group.
+        """
+        return len(self.txn_list)
+
+    def clone(self):
+        """
+        Creates a new composer with the same underlying transactions.
+        The new composer's status will be BUILDING, so additional transactions
+        may be added to it.
+        """
+        cloned = AtomicTransactionComposer()
+        cloned.method_dict = copy.deepcopy(self.method_dict)
+        cloned.txn_list = copy.deepcopy(self.txn_list)
+        for t in cloned.txn_list:
+            t.txn.group = None
+        cloned.status = AtomicTransactionComposerStatus.BUILDING
+        return cloned
+
+    def add_transaction(self, txn_and_signer):
```

Comment:
Random thought... if we want to enable a "builder pattern" type approach, this `add_XYZ` methods return `self`. So you could do something like:
```python
atc = AtomicTransactionComposer()
atc.add_transaction(... a transaction ...)
     .add_method(... a method ...)
     .add_method(... another method ...)
```

etc.

But the builder pattern is a controversial one...

Reply:
FYI, I was [conflating the Builder pattern with a Fluent Interface](https://stackoverflow.com/questions/12521712/what-is-the-pattern-used-in-java-and-elsewhere-to-set-multiple-properties-when/12521762)

Reply:
This seems like a useful improvement, since the functions currently don't return anything